runtime/stdlib/legacy_contract_upgrade_validator.go

```diff
@@ -0,0 +1,359 @@
+	for _, intf := range intersection {
+		intfs = append(intfs, validator.getInterfaceType(intf))
+	}
+	return
+}
+
+func (validator *LegacyContractUpdateValidator) requirePermitsAccess(
+	expected sema.Access,
+	found sema.EntitlementSetAccess,
+	foundType ast.Type,
+) error {
+	if !found.PermitsAccess(expected) {
+		return &AuthorizationMismatchError{
+			FoundAuthorization:    found,
+			ExpectedAuthorization: expected,
+			Range:                 ast.NewUnmeteredRangeFromPositioned(foundType),
+		}
+	}
+	return nil
+}
+
+func (validator *LegacyContractUpdateValidator) expectedAuthorizationOfComposite(composite *ast.NominalType) sema.Access {
+	// if this field is set, we are currently upgrading a formerly legacy restricted type into a reference to a composite
+	// in this case, the expected entitlements are based not on the underlying composite type,
+	// but instead the types previously in the restriction set
+	if validator.currentRestrictedTypeUpgradeRestrictions != nil {
+		return validator.expectedAuthorizationOfIntersection(validator.currentRestrictedTypeUpgradeRestrictions)
+	}
+
+	compositeType := validator.getCompositeType(composite)
+
+	if compositeType == nil {
+		return sema.UnauthorizedAccess
+	}
+
+	supportedEntitlements := compositeType.SupportedEntitlements()
+	return sema.NewAccessFromEntitlementSet(supportedEntitlements, sema.Conjunction)
+}
+
+func (validator *LegacyContractUpdateValidator) expectedAuthorizationOfIntersection(intersectionTypes []*ast.NominalType) sema.Access {
+
+	// a reference to an intersection (or restricted) type is granted entitlements based on the intersected interfaces,
+	// ignoring the legacy restricted type, as an intersection type appearing in the new contract means it must have originally
+	// been a restricted type with no legacy type
+	interfaces := validator.getIntersectedInterfaces(intersectionTypes)
+
+	supportedEntitlements := orderedmap.New[sema.EntitlementOrderedSet](0)
+
+	for _, interfaceType := range interfaces {
+		supportedEntitlements.SetAll(interfaceType.SupportedEntitlements())
+	}
+
+	return sema.NewAccessFromEntitlementSet(supportedEntitlements, sema.Conjunction)
+}
+
+func (validator *LegacyContractUpdateValidator) checkEntitlementsUpgrade(
+	oldType *ast.ReferenceType,
+	newType *ast.ReferenceType,
+) error {
+	newAuthorization := newType.Authorization
+	newEntitlementSet, isEntitlementsSet := newAuthorization.(ast.EntitlementSet)
+	foundEntitlementSet := validator.getEntitlementSetAccess(newEntitlementSet)
+
+	// if the new authorization is not an entitlements set, there's nothing to check here
+	if !isEntitlementsSet {
+		return nil
+	}
+
+	switch newReferencedType := newType.Type.(type) {
+	// a lone nominal type must be a composite
+	case *ast.NominalType:
+		expectedAccess := validator.expectedAuthorizationOfComposite(newReferencedType)
+		return validator.requirePermitsAccess(expectedAccess, foundEntitlementSet, newReferencedType)
```

Comment:
Currently this allows any set of entitlements with equal or less access permitted than the "actual" access granted by the migration. E.g. if some reference `&T` will be given `auth(E, F)` by the migration, users can update that type in a field to any of `&T`, `auth(E) &T`, `auth(F) &T`, `auth(E | F) &T` or `auth(E, F) &T`. 

Is this okay, or should we be more restrictive and require that users only migrate their field types to exactly what the migration will grant?

Reply:
@SupunS Did we address this in follow up PRs? Do we still need to discuss / address it?

Reply:
ah no, I actually missed this somehow. Created an issue: https://github.com/onflow/cadence/issues/3128